browser/src/app/LayoutingService.ts

```diff
@@ -0,0 +1,58 @@
+/* -*- js-indent-level: 8; fill-column: 100 -*- */
+
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * This file contains service which will coordinate intensive DOM modifications
+ * which may impact browser rendering performance.
+ */
+
+type LayoutingTask = () => void;
+
+class LayoutingService {
+	private _layoutTasks: Array<LayoutingTask> = [];
+	private _layoutTaskFlush: ReturnType<typeof setTimeout> | null = null;
+
+	private _flushLayoutingQueue() {
+		if (this._layoutTasks.length) {
+			window.requestAnimationFrame(() => {
+				const start = performance.now();
+				let task = this._layoutTasks.shift();
+				let duration = 0;
+				while (task) {
+					task.call(this);
+					duration = performance.now() - start;
+					if (duration > 10) {
+						this.scheduleLayouting();
+						break;
+					}
+					task = this._layoutTasks.shift();
+				}
+			});
+		}
+
+		this._layoutTaskFlush = null;
+	}
+
+	public appendLayoutingTask(task: LayoutingTask) {
+		this._layoutTasks.push(task);
```

Comment:
Is there a reason we separate scheduleLayouting from this ? seems a pain to have to remember to schedule that ;-) perhaps there is a rare case here but then scope for leakage.

Could we merge that in a follow-up ? =)

Reply:
done in https://github.com/CollaboraOnline/online/pull/11644/files